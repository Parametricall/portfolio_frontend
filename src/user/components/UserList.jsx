import React, { useState } from 'react';
import { deleteData } from '../../utilities';
import { DESTROY_USERS_URL } from '../../constants';
import { Link, useRouteMatch } from 'react-router-dom';
import { Button, Col, Container, Form, InputGroup, ListGroup, Row } from 'react-bootstrap';
import { LinkContainer } from 'react-router-bootstrap';


function UserList({fetchData}) {
  let {url} = useRouteMatch();

  const [users, setUsers] = useState(null);
  const [selected, setSelected] = useState([]);


  const getUserData = async () => {
    const userData = await fetchData();
    setUsers(userData);
  };

  React.useEffect(() => {
    getUserData();
  }, []);

  const deleteUsers = async () => {

    for (const id of selected) {
      await deleteData(`${DESTROY_USERS_URL}${id}/`)
        .catch(e => console.log(e));
    }
    setUsers(fetchData());
  };

  const userSelected = (checked, userId) => {
    if (checked) {
      selected.push(userId);
      setSelected(selected);
    } else {
      const index = selected.indexOf(userId);
      if (index !== -1) {
        selected.splice(index, 1);
        setSelected(selected);
      }
    }
  };

  let errors = null;
  let userList = null;
  if (users) {
    if (users.detail) {
      errors = <InputGroup.Text className='user-list-errors'>{users.detail}</InputGroup.Text>;
    } else {
      userList = users.map((user) => {
        return (
          <ListGroup.Item key={user.id}>
            <InputGroup>
              <InputGroup.Prepend>
                <Form.Check className="float-right" type="checkbox"
                            onChange={(e) => userSelected(e.target.checked, user.id)}/>
              </InputGroup.Prepend>
              <Link to={`${url}/${user.id}`} className='user-list-username'>{user.username}</Link>
            </InputGroup>
          </ListGroup.Item>
        );
      });
    }
  }

  return (
    <Container className="mt-5">
      <Row>
        <Col>
          <LinkContainer to={`${url}/create_user`}>
            <Button variant="primary">Create User</Button>
          </LinkContainer>
          <Button variant="danger" onClick={deleteUsers}>Delete Selected</Button>
        </Col>
      </Row>

      <Row>
        <Col>
          <ListGroup>
            {errors}
            {userList}
          </ListGroup>
        </Col>
      </Row>
    </Container>
  );
}

export default UserList;
