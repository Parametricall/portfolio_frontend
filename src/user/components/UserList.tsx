import React, { useCallback, useState } from 'react';
import { useRouteMatch } from 'react-router-dom';
import {
    Button, Col, Container, InputGroup, ListGroup, Row,
} from 'react-bootstrap';
import { LinkContainer } from 'react-router-bootstrap';
import { connect } from 'react-redux';
import UserListItem from './UserListItem';
import { DESTROY_USERS_URL, FETCH_USERS_URL } from '../../constants';
import { deleteData, fetchJsonData } from '../../utilities';
import { setUserAuthenticated } from '../../reduxStore/actions';

function UserList({ setUserAuthenticated: setUserAuthenticatedOld }) {
    const { url } = useRouteMatch();

    const [users, setUsers] = useState(null);
    const [selected, setSelected] = useState([]);

    const fetchData = useCallback(
        async () => fetchJsonData(FETCH_USERS_URL, 'GET'),
        [],
    );

    React.useEffect(() => {
        fetchData().then((userData) => setUsers(userData));
    }, [fetchData, setUserAuthenticatedOld]);

    const deleteUsers = async () => {
        // eslint-disable-next-line no-restricted-syntax
        for (const id of selected) {
            // eslint-disable-next-line no-await-in-loop
            await deleteData(`${DESTROY_USERS_URL}${id}/`).catch();
        }
        fetchData().then((userData) => setUsers(userData));
    };

    const userSelected = (checked, userId) => {
        if (checked) {
            selected.push(userId);
            setSelected(selected);
        } else {
            const index = selected.indexOf(userId);
            if (index !== -1) {
                selected.splice(index, 1);
                setSelected(selected);
            }
        }
    };

    let errors = null;
    let userList = null;
    if (users) {
        if (users.detail) {
            errors = (
                <InputGroup.Text className="user-list-errors">
                    {users.detail}
                </InputGroup.Text>
            );
        } else {
            userList = users.map((user) => (
                <UserListItem
                    key={user.id}
                    userId={user.id}
                    username={user.username}
                    onUserSelect={userSelected}
                />
            ));
        }
    }

    return (
        <Container className="mt-5">
            <Row>
                <Col>
                    <LinkContainer to={`${url}/create_user`}>
                        <Button variant="primary">Create User</Button>
                    </LinkContainer>
                    <Button variant="danger" onClick={deleteUsers}>
                        Delete Selected
                    </Button>
                </Col>
            </Row>

            <Row>
                <Col>
                    <ListGroup>
                        {errors}
                        {userList}
                    </ListGroup>
                </Col>
            </Row>
        </Container>
    );
}

export default connect(null, { setUserAuthenticated })(UserList);
