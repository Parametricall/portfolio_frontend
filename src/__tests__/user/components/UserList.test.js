import React from 'react';
import { shallow } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';

import UserList from '../../../user/components/UserList';


// https://stackoverflow.com/questions/58117890/how-to-test-components-using-new-react-router-hooks
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'), // use actual for all non-hook parts
  useRouteMatch: () => ({ url: '/users' }),
}));

describe('UserList', () => {
  const wrapper = shallow(<MemoryRouter><UserList/></MemoryRouter>, {context: ''});
  const userList = wrapper.find(UserList).dive()

  it('renders correctly', () => {
    expect(userList).toMatchSnapshot();
  });
});