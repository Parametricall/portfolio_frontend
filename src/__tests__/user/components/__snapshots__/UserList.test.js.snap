// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserList renders correctly 1`] = `
<Container
  className="mt-5"
  fluid={false}
>
  <Row
    noGutters={false}
  >
    <Col>
      <withRouter(LinkContainer)
        to="/users/create_user"
      >
        <Button
          active={false}
          disabled={false}
          variant="primary"
        >
          Create User
        </Button>
      </withRouter(LinkContainer)>
      <Button
        active={false}
        disabled={false}
        onClick={[Function]}
        variant="danger"
      >
        Delete Selected
      </Button>
    </Col>
  </Row>
  <Row
    noGutters={false}
  >
    <Col>
      <ListGroup />
    </Col>
  </Row>
</Container>
`;
